design-proposals/2022-11-29-metal-port.md

```diff
@@ -0,0 +1,426 @@
+
+Figure: Metal Renderer (general)
+
+After the Rendering Modularization rework we'll have Layers using Builders to emit Drawables.  Initially the Metal Drawables will be stubs, but we can get started with a carefully curated Metal Drawable that consists of one polygon.  It can be matched with a single Metal Shader that does nothing more than draw it.
+
+Getting the skeleton version of the Metal Renderer in place is the purpose of this work.  When it's going, the developers should see a single rectangle moving around the screen as the user pans.  Not terribly exciting, but an easy start and a good place to branch off work with multiple developers.
+
+### Textures
+
+The MapLibre Native toolkit already knows what a texture is, so this is the Metal Texture variant.  Metal textures don't deviate all that much from the OpenGL in concept, but the specifics very much do.
+
+Metal supports a whole host of texture formats that OpenGL doesn't (or does through extensions).  There are is the traditional RGBA, but there are also 32 bit float, or dual 16 bit float, or.... the list goes on and on.  We want to allow support for these without losing our minds representing them.
+
+As it stands, supporting RGBA is probably sufficient, but the utility of a good single component 16 bit texture is not to be denied and neither is the flexibility of a 32 bit floating point texture.  
+
+We should at least fill out a handful of those image types.  Which ones aren't all that important, but the workflow for adding new ones is.  It should be simple to add the logic for a new type, but we don't need to go so far as seamlessly converting from one to another.
+
+Render Targets make use of that image type enum as well, but we'll address that later.
+
+We also want a way of representing a texture that isn't ours.  That is, a texture that comes from another Metal toolkit that we may be using as input (or output).  iOS and MacOS have some incredibly powerful image processing utilities that run in real time if you know how to use them.
+
+### Off Screen Render Targets
+
+Off screen render targets are used in a couple of ways.  The most obvious is when rendering an image of a map offline.  The developer sets up the toolkit, feeds in a map and captures the result from a particular viewpoint.
+
+Less obvious is for things like heatmaps or weather data.  In those cases we want to render a tile source to a particular target, bound to a texture, and then reuse that texture later in the rendering process.  We do this kind of thing a lot with weather data in WhirlyGlobe-Maply.  You do something very similar with heatmaps.  It's a very powerful technique.
+
+The actual information around a render target is actually not that complex.  They just need to know what their format is (32 bit float?  RGBA?), how big they are and what's supposed to be drawn to them.  The rest is just an outer loop in the renderer.
```

Comment:
That's true.  I hadn't considered the stencils since we don't use them in WhirlyGlobe-Maply.  We'll need to make sure the targets can handle optional stencil buffers.

Reply:
Yeah, Mapbox tile data usually is larger than tile bound, and it uses the stencil to cut the data outside of the tile.